File: 2020/6/ruby/1.rb
require "set"

def process_part1_data(raw_data)
  raw_data.split("\n\n").map { |entry|
    entry.delete("\n").delete(" ").split("").to_set.length
  }.inject(:+)
end

def process_part2_data(raw_data)
  raw_data.split("\n\n").map { |entry|
    people = entry.split("\n").length
    values = entry.delete("\n").split("").to_set.select { |v|
      entry.count(v) == people
    }
    values.length
  }.inject(:+)
end

input_file = File.join(File.dirname(__FILE__), "../test_input")
test_data = File.read(input_file)
puts "Part 1 test result: #{process_part1_data(test_data)}"

input_file = File.join(File.dirname(__FILE__), "../input")
real_data = File.read(input_file)
puts "Part 1 result: #{process_part1_data(real_data)}"

puts "Part 2 test result: #{process_part2_data(test_data)}"
puts "Part 2 result: #{process_part2_data(real_data)}"

File: 2020/6/python/1.py
import os
from pathlib import Path


def part1(data):
    return sum(len(set(entry.replace("\n", ""))) for entry in data.split("\n\n"))


def part2(data):
    total = 0
    for entry in data.split("\n\n"):
        people = len(entry.splitlines())
        total += len(
            [
                value
                for value in set(entry.replace("\n", ""))
                if entry.count(value) == people
            ]
        )
    return total


if __name__ == "__main__":
    test_filename = Path(os.path.dirname(__file__)) / "../test_input"
    test_data = test_filename.read_text()

    filename = Path(os.path.dirname(__file__)) / "../input"
    data = filename.read_text()

    print("Test Part 1: ", part1(test_data))
    print("Part 1: ", part1(data))
    print("Test Part 2: ", part2(test_data))
    print("Part 2: ", part2(data))
